test/unit/lua_func_adapter.c

```diff
@@ -0,0 +1,267 @@
+	struct func_adapter *func = lua_func_adapter_create(L, idx);
+	func_adapter_begin(func, &ctx);
+	/* Not zero-terminated string. */
+	const char s1[] = {'a', 'b', 'c'};
+	size_t s1_len = lengthof(s1);
+	const char *s2 = "42strstr";
+	size_t s2_len = strlen(s2);
+	func_adapter_push_lstring(func, &ctx, s1, s1_len);
+	func_adapter_push_string(func, &ctx, s2);
+	int rc = func_adapter_call(func, &ctx);
+	fail_if(rc != 0);
+	ok(func_adapter_is_string(func, &ctx), "Expected string");
+	const char *retval = NULL;
+	func_adapter_pop_string(func, &ctx, &retval, NULL);
+	is(strncmp(retval, s1, s1_len), 0, "Popped string must match");
+	size_t len = 0;
+	ok(func_adapter_is_string(func, &ctx), "Expected string");
+	func_adapter_pop_string(func, &ctx, &retval, &len);
+	is(len, s1_len + s2_len, "Len does not match");
+	char *buf = tt_static_buf();
+	strncpy(buf, s1, s1_len);
+	strcpy(buf + s1_len, s2);
+	is(strcmp(retval, buf), 0, "Expected %s", buf);
+	ok(func_adapter_is_null(func, &ctx), "Expected null - no values left")
+	func_adapter_end(func, &ctx);
+	func_adapter_destroy(func);
+
+	footer();
+	check_plan();
+}
+
+static void
+test_null(void)
+{
+	plan(7);
+	header();
+
+	int idx = generate_function(
+		"function(a, b, c) return a, box.NULL, nil, c, b end");
+	size_t null_count = 4;
+	int64_t int_val = 42;
+	struct func_adapter *func = lua_func_adapter_create(L, idx);
+	func_adapter_begin(func, &ctx);
+	func_adapter_push_null(func, &ctx);
+	func_adapter_push_integer(func, &ctx, int_val);
+	int rc = func_adapter_call(func, &ctx);
+	fail_if(rc != 0);
+	for (size_t i = 0; i < null_count; ++i) {
+		ok(func_adapter_is_null(func, &ctx), "Expected null")
+		func_adapter_pop_null(func, &ctx);
+	}
+	ok(func_adapter_is_integer(func, &ctx), "Expected integer");
+	int64_t integer_retval = 0;
+	func_adapter_pop_integer(func, &ctx, &integer_retval);
+	ok(func_adapter_is_null(func, &ctx), "Expected null - no values left")
+	func_adapter_end(func, &ctx);
+	func_adapter_destroy(func);
+
+	is(integer_retval, int_val, "Returned integer must be as expected");
+
+	footer();
+	check_plan();
+}
+
+static void
+test_error(void)
+{
+	plan(2);
+	header();
+
+	const char *functions[] = {
+		"function() error('lua error') end",
+		"function() box.error('tnt error') end"
+	};
+
+	for (size_t i = 0; i < lengthof(functions); ++i) {
+		int idx = generate_function(functions[i]);
+		struct func_adapter *func = lua_func_adapter_create(L, idx);
+		func_adapter_begin(func, &ctx);
+		int rc = func_adapter_call(func, &ctx);
+		is(rc, -1, "Call must fail");
+		func_adapter_end(func, &ctx);
+		func_adapter_destroy(func);
+	}
+
+	footer();
+	check_plan();
+}
+
+static int
+test_lua_func_adapter(void)
+{
+	plan(5);
+	header();
+
+	test_numeric();
+	test_tuple();
+	test_string();
+	test_null();
+	test_error();
+
+	footer();
+	return check_plan();
+}
+
+int
+main(void)
+{
+	memory_init();
+	fiber_init(fiber_c_invoke);
+	tuple_init(NULL);
+
+	L = luaT_newteststate();
+	tarantool_L = L;
```

Comment:
Nit: There's no need to make `L` global if you set `tarantool_L` to it anyway. Please use `tarantool_L` instead of `L`.

Reply:
Done.